tests/lib/ocpp/v201/test_smart_charging_handler.cpp

```diff
@@ -147,11 +169,35 @@ class ChargepointTestFixtureV201 : public testing::Test {
         };
     }
 
-    DeviceModel create_device_model() {
-        std::unique_ptr<DeviceModelStorageMock> storage_mock =
-            std::make_unique<testing::NiceMock<DeviceModelStorageMock>>();
-        ON_CALL(*storage_mock, get_device_model).WillByDefault(testing::Return(DeviceModelMap()));
-        return DeviceModel(std::move(storage_mock));
+    void create_device_model_db(const std::string& path) {
+        sqlite3* source_handle;
+        sqlite3_open(DEVICE_MODEL_DB_LOCATION_V201, &source_handle);
+        sqlite3_open(path.c_str(), &db_handle);
+
+        auto* backup = sqlite3_backup_init(db_handle, "main", source_handle, "main");
+        sqlite3_backup_step(backup, -1);
+        sqlite3_backup_finish(backup);
+
+        sqlite3_close(source_handle);
+    }
+
+    std::shared_ptr<DeviceModel>
+    create_device_model(const std::string& path = "file:device_model?mode=memory&cache=shared",
```

Comment:
I've created a PR here: https://github.com/EVerest/libocpp/pull/664

Let me know if this would be usable for you or not. I had to change to a `shared_ptr` for the `EvseInterface` but I have an idea on how to remove the need for passing the map by reference as well, eliminating the `shared_ptr` too.

Reply:
Ultimately I think that would be great, but the issue we ran into was that `get_value()`, `get_optional_value()`, and `request_value()` return templated types, and I could not find a way to mock those.

How would I request a variable using that PR?

Reply:
I found that issue as well and had to work around it a little bit. If you want to mock the template function `get_value()` for example, you could use `expect_request_value()` to set the return value.

This is possible since all the templated functions only use a call to `request_value_internal()` that is mockable. `expect_request_value()` is just a helper to simplify the usage a bit.

there are probably a lot of things that can still be improved or added but it should be a good starting point I think.

Reply:
For example this should do the trick:
```
this->device_model.expect_request_value(ControllerComponentVariables::ChargingStationSupplyPhases, "5");

EXPECT_EQ(this->device_model.get_value<int32_t>(ControllerComponentVariables::ChargingStationSupplyPhases), 5);
```

Reply:
Talking this through with the team, I think we'd prefer to keep it as is. Using the real database and loading it in-memory gives us free testing of real state, and gives us code that's easier to parse than using a mock.